src/workerd/jsg/wrappable.c++

```diff
@@ -14,37 +14,114 @@ namespace workerd::jsg {
 
 class Wrappable::CppgcShim final: public cppgc::GarbageCollected<CppgcShim> {
 public:
-  CppgcShim(Wrappable& wrappable): wrappable(kj::addRef(wrappable)) {
+  CppgcShim(Wrappable& wrappable): state(Active { kj::addRef(wrappable) }) {
     KJ_DASSERT(wrappable.cppgcShim == nullptr);
     wrappable.cppgcShim = *this;
   }
 
   ~CppgcShim() {
-    KJ_IF_MAYBE(w, wrappable) {
-      KJ_DASSERT(&KJ_ASSERT_NONNULL(w->get()->cppgcShim) == this);
-      KJ_DASSERT(w->get()->strongWrapper.IsEmpty());
-      w->get()->detachWrapper();
+    KJ_SWITCH_ONEOF(state) {
+      KJ_CASE_ONEOF(active, Active) {
+        KJ_DASSERT(&KJ_ASSERT_NONNULL(active.wrappable->cppgcShim) == this);
+        KJ_DASSERT(active.wrappable->strongWrapper.IsEmpty());
+        active.wrappable->detachWrapper(false);
+      }
+      KJ_CASE_ONEOF(freelisted, Freelisted) {
+        KJ_DASSERT(&KJ_ASSERT_NONNULL(*freelisted.prev) == this);
+        *freelisted.prev = freelisted.next;
+        KJ_IF_MAYBE(next, freelisted.next) {
+          KJ_DASSERT(next->state.get<Freelisted>().prev == &freelisted.next);
+          next->state.get<Freelisted>().prev = freelisted.prev;
+        }
+      }
+      KJ_CASE_ONEOF(d, Dead) {}
     }
   }
 
   void Trace(cppgc::Visitor* visitor) const {
-    KJ_IF_MAYBE(w, wrappable) {
-      w->get()->traceFromV8(*visitor);
+    KJ_SWITCH_ONEOF(state) {
+      KJ_CASE_ONEOF(active, Active) {
+        active.wrappable->traceFromV8(*visitor);
+      }
+      KJ_CASE_ONEOF(freelisted, Freelisted) {
+        // We're tracing a shim for an object that was collected in minor GC. This could happen
+        // due to conservative GC or due to incremental marking. Unfortunately the shim won't be
+        // collected on this pass but hopefully it can be on the next pass.
+      }
+      KJ_CASE_ONEOF(d, Dead) {}
     }
   }
 
-  mutable kj::Maybe<kj::Own<Wrappable>> wrappable;
-  // This can become null if the wrappable is force-collected without waiting for GC.
+  struct Active {
+    kj::Own<Wrappable> wrappable;
+  };
+  struct Freelisted {
+    // The JavaScript wrapper using this shim was collected in a minor GC. cppgc objects can only
+    // be collected in full GC, so we freelist the shim object in the meantime.
+
+    kj::Maybe<Wrappable::CppgcShim&> next;
+    kj::Maybe<Wrappable::CppgcShim&>* prev;
+    // kj::List doesn't quite work here because the list link is inside a OneOf. Also we want a
+    // FIFO list anyway so we don't need a tail pointer, which makes things easier. So we do it
```

Comment:
`allocateShim()` pops and returns the head of the list (if non-null), while `addToFreelist(shim)` always sets the head to `shim`. Doesn't that make this list LIFO, not FIFO?

Do we actually want FIFO?

Reply:
I meant to say LIFO here.